packages/ember-data/lib/serializers/json-api-serializer.js

```diff
@@ -0,0 +1,294 @@
+
+    if (!data) { return; }
+
+    typeName = this.normalizeTypeKey(data.type);
+
+    Ember.assert('No model was found for model name "' + typeName + '"', store.modelFactoryFor(typeName));
+
+    type = store.modelFor(typeName);
+    typeSerializer = store.serializerFor(type);
+
+    return typeSerializer.normalize(type, data);
+  },
+
+  extractIncluded: function(store, included) {
+    var type, typeName, typeSerializer, hash;
+
+    if (!included) { return; }
+
+    forEach.call(included, function(data) {
+      typeName = this.normalizeTypeKey(data.type);
+
+      if (!store.modelFactoryFor(typeName)) {
+        Ember.warn('No model was found for model name "' + typeName + '"', false);
+        return;
+      }
+
+      type = store.modelFor(typeName);
+      typeSerializer = store.serializerFor(type);
+
+      hash = typeSerializer.normalize(type, data);
+      store.push(typeName, hash);
+    }, this);
+  },
+
+
+
+  serialize: function(snapshot, options) {
+    var json = {};
+
+    json['type'] = this.serializeTypeKey(snapshot.typeKey);
+
+    if (options && options.includeId) {
+      json['id'] = snapshot.id;
+    }
+
+    snapshot.eachAttribute(function(key, attribute) {
+      this.serializeAttribute(snapshot, json, key, attribute);
+    }, this);
+
+    snapshot.eachRelationship(function(key, relationship) {
+      if (relationship.kind === 'belongsTo') {
+        this.serializeBelongsTo(snapshot, json, relationship);
+      } else if (relationship.kind === 'hasMany') {
+        this.serializeHasMany(snapshot, json, relationship);
+      }
+    }, this);
+
+    json = { data: json };
+
+    return json;
+  },
+
+  serializeAttribute: function(snapshot, json, key, attribute) {
+    var value = snapshot.attr(key);
+    var type = attribute.type;
+
+    if (type) {
+      var transform = this.transformFor(type);
+      value = transform.serialize(value);
+    }
+
+    var payloadKey = this.keyForAttribute(key);
+    json[payloadKey] = value;
+  },
+
+  serializeBelongsTo: function(snapshot, json, relationship) {
+    var key = relationship.key;
+    var belongsTo = snapshot.belongsTo(key);
+
+    var links = json['links'] = json['links'] || {};
+
+    var payloadKey = this.keyForRelationship(key, 'belongsTo');
+
+    if (Ember.isNone(belongsTo)) {
+      links[payloadKey] = { id: null };
+    } else {
+      links[payloadKey] = {
+        type: this.serializeTypeKey(belongsTo.typeKey),
+        id: belongsTo.id
+      };
+    }
+  },
+
+  serializeHasMany: function(snapshot, json, relationship) {
+    var key = relationship.key;
+    var hasMany = snapshot.hasMany(key);
+
+    var links = json['links'] = json['links'] || {};
+
+    var payloadKey = this.keyForRelationship(key, 'hasMany');
+
+    if (hasMany.length === 0) {
+      links[payloadKey] = { ids: [] };
+    } else {
+
+      // TODO: if all items in hasMany is of the same type, provide
+      // { type: "type", ids: [..] } instead of { data: [...] }
```

Comment:
I don't think this `TODO` is valid as of https://github.com/json-api/json-api/pull/447. Additionally, the key is `linkage` now, instead of `data`